src/client/opamSolution.ml

```diff
@@ -474,6 +478,79 @@ let parallel_apply t
     else action_graph
   in
 
+  (* factorise fetches: creates [`Fetches] nodes that download once the archive
+     and copies its content in source directories. Handles only http and rsync
+     backend with at least a checksum. *)
+  let action_graph =
+    let cache_dir = OpamRepositoryPath.download_cache t.switch_global.root in
+    let packages =
+      let urls, hash_nvs =
+        OpamPackage.Set.fold (fun nv (urls, hashes as acc) ->
+            match OpamSwitchState.url t nv with
+            | Some url ->
+              let checksums = OpamFile.URL.checksum url in
+              let url = OpamFile.URL.url url in
+              (match url.OpamUrl.backend, checksums with
+               | (`http | `rsync ) , _::_
+                 when OpamFilename.is_archive
+                     (OpamFilename.of_string url.OpamUrl.path) ->
+                 OpamUrl.Map.update url
+                   (fun hset -> OpamHash.Set.(union hset (of_list checksums)))
+                   OpamHash.Set.empty urls,
+                 List.fold_left (fun hashes h ->
+                     OpamHash.Map.update h (OpamPackage.Set.add nv)
+                       OpamPackage.Set.empty hashes)
+                   hashes checksums;
+               | _ -> acc)
+            | None -> acc)
+          sources_needed (OpamUrl.Map.empty, OpamHash.Map.empty)
+      in
+      OpamUrl.Map.fold (fun _url hashes packages ->
+          (* should we check urls ? *)
+          let _, choosen =
+            (* should we keep other checksums or drop them ? *)
```

Comment:
The general rule is that caching/indexing may behave differently, but all of them are verified when checking. We're not allowed to drop!
This raises a question when there are multiple sources with different hashes: here we will now have to merge all hashes, so from a set of packages with the same source, adding a wrong hash to one of them could "contaminate" the others and prevent installation for all of them. It seems alright though, I don't see how we could factorise without doing it like this.

Reply:
In fact there was something wrong: only the url of the first/chosen package is kept, so if another package define another checksums, they are not checked.
We need to check all checksums for each package: after downloading the archive, before copying its content in source directory, check package defined checksums, and fail otherwise.

Reply:
yes definitely 2. ; it's mitigated by the fact that it's only true for the package you are trying to install in the current run.